crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -192,29 +220,42 @@ impl<'db> NarrowingConstraintsBuilder<'db> {
             {
                 // SAFETY: we should always have a symbol for every Name node.
                 let symbol = self.symbols().symbol_id_by_name(id).unwrap();
-                let comp_ty = inference.expression_ty(right.scoped_ast_id(self.db, scope));
-                match op {
-                    ast::CmpOp::IsNot => {
-                        if comp_ty.is_singleton() {
+                let rhs_ty = inference.expression_ty(right.scoped_ast_id(self.db, scope));
+                match (op, negative) {
+                    (ast::CmpOp::IsNot, false) | (ast::CmpOp::Is, true) => {
+                        if rhs_ty.is_singleton() {
                             let ty = IntersectionBuilder::new(self.db)
-                                .add_negative(comp_ty)
+                                .add_negative(rhs_ty)
                                 .build();
                             self.constraints.insert(symbol, ty);
                         } else {
                             // Non-singletons cannot be safely narrowed using `is not`
                         }
                     }
-                    ast::CmpOp::Is => {
-                        self.constraints.insert(symbol, comp_ty);
+                    (ast::CmpOp::Is, false) | (ast::CmpOp::IsNot, true) => {
+                        self.constraints.insert(symbol, rhs_ty);
                     }
-                    ast::CmpOp::NotEq => {
-                        if comp_ty.is_single_valued(self.db) {
+                    (ast::CmpOp::NotEq, false) | (ast::CmpOp::Eq, true) => {
+                        if rhs_ty.is_single_valued(self.db) {
                             let ty = IntersectionBuilder::new(self.db)
-                                .add_negative(comp_ty)
+                                .add_negative(rhs_ty)
                                 .build();
                             self.constraints.insert(symbol, ty);
                         }
                     }
+                    (ast::CmpOp::NotEq, true) | (ast::CmpOp::Eq, false) => {
+                        if rhs_ty.is_single_valued(self.db) {
+                            self.constraints.insert(symbol, rhs_ty);
+                        } else if let Some(union) = rhs_ty.into_union_type() {
+                            if union
+                                .elements(self.db)
+                                .iter()
+                                .all(|ty| ty.is_single_valued(self.db))
+                            {
+                                self.constraints.insert(symbol, rhs_ty);
+                            }
+                        }
+                    }
```

Comment:
As discussed in another comment in the tests, I don't think we can generally narrow on equality checks (or negative inequality checks). But I think we can (and probably will need to) do a more limited form of this, where we check the type of the LHS and if it is a union of single-valued types, we eliminate (by constraining with a negative intersection of) any single-valued types that we know are disjoint with the RHS.

In other words, while we can't infer that `x: int; if x == 1:` results in `Literal[1]`, we can infer that `x: Literal[1, 2, 3]; if x == 1:` results in `Literal[1]`, and I think we'll need to be able to make the latter inference.

(There may be a more general way to frame this than what I wrote above, need to think about it.)

Reply:
This represents a more general category of narrowing cases that we will also run into with, for example, simple `if not x:` narrowing, where the information gained from the narrowing in that case is "`x` is falsy" and in this case is "`x` is equal to 1", neither of which is a constraint that can correctly be expressed simply a type to intersect `x` with, given our currently available types. I spent some more time thinking about this, and wrote it up at https://github.com/astral-sh/ruff/issues/13694#issuecomment-2438438759

I'm afraid that's a bit long, but I think the relevant conclusion for this PR is, let's not handle it in this PR :) For this PR, let's just remove the narrowing on `==` (or negated `!=`) entirely for now.